### PR TITLE
Add observability check to counters

### PR DESCRIPTION
**Pull request by AlexPadron**
This is for issue https://github.com/prometheus/client_python/issues/454

### CODE REVIEW

File: prometheus_client/metrics.py [hunk 1]
```diff
@@ -58,6 +58,16 @@ def _is_observable(self):
         # * the child of a labelled metric.
         return not self._labelnames or (self._labelnames and self._labelvalues)
 
+    def _raise_if_not_observable(self):
+        # Functions that mutate the state of the metric, for example incrementing
+        # a counter, will fail if the metric is not observable, because only if a
+        # metric is observable will the value be initialized.
+        if not self._is_observable():
+            raise ValueError(
+                '%s metric is not observable. This is likely because the metric requires '
```
Comment (AlexPadron) [inline prometheus_client/metrics.py]: Thinking if we want to make this more detailed, for example we could use inspect and get the method name that is being invoked. Also, we may want to use this in other metrics as well. I added the check to counters, but the same issue exists in gauges as well for example

File: prometheus_client/metrics.py [hunk 2]
```diff
@@ -239,8 +249,11 @@ def _metric_init(self):
 
     def inc(self, amount=1):
         """Increment counter by the given amount."""
+        self._raise_if_not_observable()
```
Comment (brian-brazil) [inline prometheus_client/metrics.py]: This is a hot path, we shouldn't add code here.

File: prometheus_client/metrics.py (hunk 1 above)
Comment (brian-brazil) [reply prometheus_client/metrics.py]: I'm not sure "observable" is the right word to use here, it'll confuse users. Missing label values would be more accurate.

**Comment by brian-brazil** [comment]
Other metrics types also need this, e.g. the time() functions.

**Comment by AlexPadron** [comment]
> Other metrics types also need this, e.g. the time() functions.

I think I got all of these (at least in metrics.py). As far as I can tell by grepping this is the only location where the metrics are defined

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**